feat: add more v1 methods to v2 client

sirewix opened the pull request.

tlater-famedly reviewed (commented):
Nice, but I'm not a big fan of sneaking in a bunch of refactoring with a feature commit.

I avoid being *too* precise with this because it can be a ton of work for more hairy commits, but in this case it should be really trivial. Can we split this into two commits, one with the refactor and another with the new method?

tlater-famedly commented on src/v2/management/mod.rs:
```diff
@@ -213,13 +213,31 @@ impl Zitadel {
 	}
+
+	/// List Project Grant Members [Docs](https://zitadel.com/docs/apis/resources/mgmt/management-service-list-project-grant-members)
+	pub fn list_project_grant_members(
```
We should probably also add a test case for this.

sirewix commented:
I agree, initially I added the support for org id in headers, which required changing all the lines where `PaginationHandler` was used, so I decided it was a good chance to add one type parameter. And then I rebased my branch and turned out the org id feature was added already

tlater-famedly commented:
We should try to add proper sorting to these methods upstream.

sirewix commented on tests/e2e.rs:
```diff
@@ -99,6 +114,305 @@ async fn test_e2e_create_token() -> Result<()> {
+#[test(tokio::test)]
+#[test_log(default_log_filter = "debug")]
+#[allow(trivial_casts)]
+async fn test_e2e_pupa_lupa() -> Result<()> {
```
Any suggestions for a good name?

tlater-famedly replied:
`test_role_methods`?

tlater-famedly commented on src/v2/management/models/management_service_list_project_grant_members_body.rs:
```diff
@@ -69,3 +70,19 @@ impl ManagementServiceListProjectGrantMembersBody {
 	}
 }
+
+impl crate::v2::pagination::PaginationRequest for Option<Vec<models::Zitadelmemberv1SearchQuery>> {
```
I'd prefer adding this to `src/v2/management/mod.rs`, since this is deep in the generated code; editing that as little as possible makes it easier to re-generate if/when that becomes necessary.

We're in the same crate, so adding `impl`s elsewhere is luckily permitted :D

sirewix replied:
An argument against this is that mod.rs would get bloated, and since this impl is for a particular type we should put it near that type to keep things neat and clean.

An argument for this is that this impl is used in several requests (they technically have different request types but they are identical).

sirewix replied:
Moved to `v2/management/mod.rs`

tlater-famedly commented on src/v2/management/models/zitadelmemberv1_search_query.rs:
```diff
@@ -19,12 +19,16 @@ use crate::v2::management::models;
 #[derive(Clone, Debug, Serialize, Deserialize)]
 pub struct Zitadelmemberv1SearchQuery {
 	#[serde(rename = "firstNameQuery")]
+	#[serde(skip_serializing_if = "Option::is_none")]
```
Does Zitadel not support `none` for these?

sirewix replied:
It was me experimenting with something, I'll remove it

tlater-famedly commented on src/v2/management/models/v1_user_grant_query.rs:
```diff
@@ -10,353 +10,86 @@
-	pub fn with_project_id_query(
-		mut self,
+#[serde(untagged)]
+pub enum V1UserGrantQuery {
```
I get it, the codegen isn't exactly ideal, buuuut, now we're not really getting its benefits either. I'd be on board with hand-rolling all types if we want to do that as a separate piece of work, but I'm not fond of doing a mix of the two, this kind of gives us the worst of both worlds.

It's really a shame that grpc is so finnicky with all the proxies; working directly with the protobuf definitions would make a lot of this significantly less messy (not *perfect*, but at least as perfect as you're going to get when limited by grpc types). They're also more accurate since Zitadel themselves use them and don't go through the clearly half-baked rust implementation of openapi's codegen.

tlater-famedly replied:
Yeah, I see your point. We should probably think about a longer-term solution to these things; the codegen is anyway suboptimal most of the time, and clearly Zitadel doesn't document their API well enough for proper codegen anyway.

But that's besides the point of this PR, and I think hand-rolling this type is reasonable. It'd be my preference to do that outside of the generated modules, so I probably would do this with a separate struct and `impl From<>`, but we're already editing the generated code enough that that ship has sailed.

tlater-famedly commented on tests/e2e.rs:
```diff
@@ -99,6 +114,305 @@ async fn test_e2e_create_token() -> Result<()> {
+
+	let roles = |role: &str| vec![role.to_owned()];
+
+	// org 1
```
Could we use a bit of control flow to make this more readable, or at least add some more comments? I think I get the gist of it, you're creating some org/project/user structure and assert that it's being listed correctly, but this is really hard to read.

I'm also not sure we need this complex of a test, ultimately we just want to assert that we're calling the API correctly and process the response correctly; if Zitadel misbehaves that's not really our problem.

sirewix replied:
> but this is really hard to read.

I don't think theres a way to make it easy to read. The structure is intentionally complex to cover as much cases as possible

> I'm also not sure we need this complex of a test, ultimately we just want to assert that we're calling the API correctly and process the response correctly; if Zitadel misbehaves that's not really our problem.

It's not to test one or another method, it's more to actually figure out how it works. Think of these assertions as documentation how it works and what the zitadel is actually returning (I wish they had that just documented somewhere, so we wouldn't need to experiment with it ourselves)

tlater-famedly reviewed (commented):
Love the unboxing of the `PaginationHandler`, thanks for taking your time to do that.

tlater-famedly commented on tests/e2e.rs:
```diff
@@ -99,6 +168,268 @@ async fn test_e2e_create_token() -> Result<()> {
+	Zitadel::new(url, service_account_file.to_path_buf()).await
+}
+
+#[test(tokio::test)]
```
We *could* consider putting this in a separate file.

Your call, won't block the PR on this.

sirewix replied:
We could, but it would require major refactoring of our tests, ideally moving other tests into other modules.

tlater-famedly reviewed (changes requested):
Ready to go except for one detail that almost slipped by me?

sirewix commented on .gitignore:
```diff
@@ -1,5 +1,7 @@
 /target
 /cargo_home
+/tests/environment/zitadel/service-user.json
+/Cargo.lock
```
Yes, it is. Dependencies for libraries should not be locked, otherwise downstream crates will have a high chance of dependency conflicts. We can factor out this fix into another PR, but I don't really think it's worth it, as existing ZRC users have their own lock files so it should not cause any immedite conflicts

https://doc.rust-lang.org/1.71.1/cargo/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries

tlater-famedly replied:
That is outdated, hasn't been the case since 2023: https://doc.rust-lang.org/stable/cargo/faq.html#why-have-cargolock-in-version-control

Let's do this in a separate PR

sirewix replied:
> Cargo.lock does not affect the consumers of your package, only Cargo.toml does that. 

Ah, alright, let me revert this

sirewix replied:
Done

tlater-famedly commented on src/v2/management/mod.rs:
```diff
@@ -226,11 +226,12 @@ impl Zitadel {
 		org_id: Option<String>,
 		project_id: &str,
-		queries: Vec<Zitadelmemberv1SearchQuery>,
+		params: Option<PaginationParams>,
```
This is a breaking change, the commit should be marked as such.

tlater-famedly replied:
Maybe best done as part of #57 ?

sirewix replied:
It's not breaking in the scope of this PR. Compared to previous commit, yes, but there's not going to be a release in between. Honestly I'd just squash all commits in this PR into one. Another option would be to put these changes into first commit, but such pedantic approach just takes too much time to be reasonable imo.

tlater-famedly replied:
Ah, I overlooked that, sorry. Yeah, frankly not sure how to best handle this in general; while it doesn't cause breakage on a release level, if we go by conventional commits it *should* be marked as a breaking change, since the history isn't reverted PR-by-PR but commit-by-commit.

Squashing d9b5213c4aece2b7c3e417a848745a76f5000e07 -> a846e9db9583777c9bc2976c5346f35af4379e8e makes sense to me, and would solve this.
